src/users.rs

```
@@ -0,0 +1,243 @@
+const N_ITER: u32 = 100_000;
+
+// Error Used for User related Functions
+// Implements the traits of Error
+#[derive(Debug)]
+struct UserError(String);
+impl fmt::Display for UserError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+impl Error for UserError {}
+
+// Generates a salt
+fn get_salt() -> Vec<u8> {
+    let salt: Vec<u8> = (0..CREDENTIAL_LEN).map(|_| rand::random::<u8>()).collect();
+    salt
+}
+
+// Takes in a salt and password and returns the hash
+// using the pbkdf2 algorithm
+fn get_hash_from_password(password: String, salt: Vec<u8>) -> Vec<u8> {
+    let mut pbkdf2_hash = [0u8; CREDENTIAL_LEN];
+    pbkdf2::derive(
+        &digest::SHA512,
+        N_ITER,
+        &salt,
+        password.as_bytes(),
+        &mut pbkdf2_hash,
+    );
+    pbkdf2_hash.to_vec()
+}
+
+// Verifies a password and salt against a hash
+// using the pbkdf2 algorithm
+fn verify_password(
+    password: String,
+    salt: Vec<u8>,
+    pbkdf2_hash: Vec<u8>,
+) -> Result<(), Unspecified> {
+    pbkdf2::verify(
+        &digest::SHA512,
+        N_ITER,
+        &salt,
+        password.as_bytes(),
+        &pbkdf2_hash,
+    )
+}
+
+// creates a sqlite DB if it does not exist and initializes the user table
+// returns the connection
+fn connection() -> Result<rusqlite::Connection, Box<dyn Error>> {
+    let conn = Connection::open("sqlite.db").expect("db conn fail");
+
+    // create the users table if it does not exist
+    match conn.execute(
+        "create table if not exists users (
+             id integer primary key,
+             name text not null unique,
+             salt text not null,
+             hash text not null,
+             active integer default 1 not null
+         );",
+        NO_PARAMS,
+    ) {
+        Ok(_) => {
+            return Ok(conn);
+        }
+        Err(_) => {
+            return Err(Box::new(UserError("Error Creating table".into())));
```

Is there a reason (e.g. security) you don't want to return the original error? Then you can do:

```
conn.execute(...)?;
Ok(conn)
```

---

My thinking was that there are a bunch of different errors that can be returned just normalize them all to one type of error, but I will do your suggestion above